ssl/ssl_lib.c

```diff
@@ -4118,6 +4171,43 @@ SSL_CTX *SSL_CTX_new_ex(OSSL_LIB_CTX *libctx, const char *propq,
         goto err;
     }
 
+#ifndef OPENSSL_NO_SSLKEYLOG
+    if (keylogfile != NULL && strlen(keylogfile) != 0) {
+        /* Make sure we have a global lock allocated */
+        if (!RUN_ONCE(&ssl_keylog_once, ssl_keylog_init)) {
+            /* log an error, but don't fail here */
+            ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
+            goto err;
+        }
+
+        /* Grab out global lock */
+        if (!CRYPTO_THREAD_write_lock(keylog_lock)) {
+            ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
+        } else {
+            /*
+             * If the bio for the requested keylog file hasn't been
+             * created yet, go ahead and create it, and set it to append
+             * if its already there.
+             */
+            if (keylog_bio == NULL) {
+                keylog_bio = BIO_new_file(keylogfile, "a");
+                if (keylog_bio == NULL)
+                    ERR_raise(ERR_LIB_SSL, ERR_R_SSL_LIB);
```

Comment:
This is weird. We raise an error, but then carry on regardless and still return `ret` which is non-NULL and therefore the function will appear to succeed in this case.

Reply:
I agree it is wierd, but I thought it was the right approach.  I.e. if an SSL_CTX is successfully created, but then we fail creating the keylog file, do we want to fail the whole function?  I can certainly change that, but I thought it might be better, as this is a debug facility, to allow the ssl context to get successfuly created, and just warn about the lack of log file with an error on the stack

Reply:
I think this is wrong. Basically we should never put an error on the error stack but still return success. The hint to go look in the error stack is the fact that we fail the call. Without this its likely this stale error will go unnoticed and (might) cause problems later.

Reply:
I think this should be still reconsidered. Because currently we let `SSL_CTX_new_ex()` to fail whenever `SSLKEYLOG` file can not be created. For example user may do some typo and variable points to location where such file can not be created due to lack of permissions. We basically turn kind of innocent failure to hard failure which may be significantly surprising for consumer (application). For example think of application which fails to connect to server just because SSLKEYLOG file can not be created. This can happen suddenly and leaves user with little clue. Keeping application working but no keys being logged allows user to stay focused on what happens with keylog because things do work but no keys are logged. This is different story to 'nothing works now'.

I would prefer to create trace/warning message and let openssl to return newly created context.

Reply:
The `ERR_raise` API is *not* a warning mechanism. The closest we have to such a thing is the TRACE API. My objection here is that we are misusing `ERR_raise`. I could live with a decision to ignore failures here and carry on regardless - if that does not result in an error going on the error stack. We must never raise an error and return success.